From 3e98a54ea525b6f904592b01ed3cf0d21c0a22c9 Mon Sep 17 00:00:00 2001
From: Brent Cook <busterb@gmail.com>
Date: Thu, 1 Jan 2015 07:18:11 -0600
Subject: [PATCH 06/25] allow overriding default user and file locations

Allow the build process to override the default ntpd file paths and
default user. On some systems, runstatedir can be a tmpfs, so we need to
create the parent path on start there.
---
 src/usr.sbin/ntpd/ntpd.c |  4 ++++
 src/usr.sbin/ntpd/ntpd.h | 13 ++++++++++---
 2 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/src/usr.sbin/ntpd/ntpd.c b/src/usr.sbin/ntpd/ntpd.c
index e70c87f25..3e7475efe 100644
--- a/src/usr.sbin/ntpd/ntpd.c
+++ b/src/usr.sbin/ntpd/ntpd.c
@@ -20,6 +20,7 @@
 #include <sys/types.h>
 #include <sys/resource.h>
 #include <sys/socket.h>
+#include <sys/stat.h>
 #include <sys/sysctl.h>
 #include <sys/wait.h>
 #include <sys/un.h>
@@ -261,6 +262,9 @@ main(int argc, char *argv[])
 	if (chdir("/") == -1)
 		fatal("chdir(\"/\")");
 
+	if (mkdir(RUNSTATEDIR, 0755) == -1 && errno != EEXIST)
+		fatal("mkdir %s", RUNSTATEDIR);
+
 	signal(SIGCHLD, sighdlr);
 
 	/* fork child process */
diff --git a/src/usr.sbin/ntpd/ntpd.h b/src/usr.sbin/ntpd/ntpd.h
index 895581696..4a3694428 100644
--- a/src/usr.sbin/ntpd/ntpd.h
+++ b/src/usr.sbin/ntpd/ntpd.h
@@ -36,10 +36,17 @@
 
 #define MAXIMUM(a, b)	((a) > (b) ? (a) : (b))
 
+#ifndef NTPD_USER
 #define	NTPD_USER	"_ntp"
-#define	CONFFILE	"/etc/ntpd.conf"
-#define DRIFTFILE	"/var/db/ntpd.drift"
-#define	CTLSOCKET	"/var/run/ntpd.sock"
+#endif
+
+#ifndef	SYSCONFDIR
+#define	SYSCONFDIR	"/etc"
+#endif
+#define	CONFFILE	SYSCONFDIR "/ntpd.conf"
+
+#define	DRIFTFILE	LOCALSTATEDIR "/ntpd.drift"
+#define	CTLSOCKET	RUNSTATEDIR "/ntpd.sock"
 
 #define	INTERVAL_QUERY_NORMAL		30	/* sync to peers every n secs */
 #define	INTERVAL_QUERY_PATHETIC		60
-- 
2.55.0

